dallinger/config.py

```diff
@@ -0,0 +1,163 @@
+        self.extend(data, cast_types=True)
+
+    def load_from_environment(self):
+        self.extend(os.environ, cast_types=True)
+
+    def load_config(self):
+        if self.ready:
+            raise ValueError("Already loaded")
+        globalConfigName = ".dallingerconfig"
+        localConfigName = "config.txt"
+        if 'OPENSHIFT_SECRET_TOKEN' in os.environ:
+            globalConfig = os.path.join(os.environ['OPENSHIFT_DATA_DIR'], globalConfigName)
+        else:
+            globalConfig = os.path.expanduser(os.path.join("~/", globalConfigName))
+        localConfig = os.path.join(os.getcwd(), localConfigName)
+
+        defaults_folder = os.path.join(os.path.dirname(__file__), "default_configs")
+        local_defaults_file = os.path.join(defaults_folder, "local_config_defaults.txt")
+        global_defaults_file = os.path.join(defaults_folder, "global_config_defaults.txt")
+        if not os.path.exists(localConfig):
+            raise ValueError("No config.txt in the current directory")
+
+        # read default global and local, then user's global and local. This way
+        # any field not in the user's files will be set to the default value.
+        for config_file in [global_defaults_file, local_defaults_file, globalConfig, localConfig]:
+            self.load_from_config_file(config_file)
+        self.ready = True
+
+
+configurations = threading.local()
+
+
+def get_config():
+    if hasattr(configurations, 'config'):
+        return configurations.config
+    configurations.config = Configuration()
+
+    for key, type_, synonyms in (
+        ('ad_group', unicode, []),
+        ('amt_keywords', unicode, []),
+        ('anonymize_data', bool, []),
+        ('approve_requirement', int, []),
+        ('auto_recruit', bool, []),
+        ('aws_access_key_id', unicode, []),
+        ('aws_region', unicode, []),
+        ('aws_secret_access_key', unicode, []),
+        ('base_payment', float, []),
+        ('browser_exclude_rule', unicode, []),
+        ('clock_on', bool, []),
+        ('contact_email_on_error', unicode, []),
+        ('database_size', unicode, []),
+        ('database_url', unicode, []),
+        ('description', unicode, []),
+        ('duration', float, []),
+        ('dyno_type', unicode, []),
+        ('host', unicode, ['HOST']),
+        ('port', int, ['PORT']),
+        ('launch_in_sandbox_mode', bool, []),
+        ('lifetime', int, []),
+        ('logfile', unicode, []),
+        ('loglevel', int, []),
+        ('mode', unicode, []),
+        ('notification_url', unicode, []),
+        ('num_dynos_web', int, []),
+        ('num_dynos_worker', int, []),
+        ('num_participants', int, []),
+        ('organization_name', unicode, []),
+        ('psiturk_keywords', unicode, []),
+        ('table_name', unicode, []),
+        ('threads', unicode, []),
+        ('title', unicode, []),
+        ('us_only', bool, [])
+    ):
+        configurations.config.register(key, type_, synonyms)
```

Comment:
My thinking here was that experiments need to be able to register their own configuration variables, so they need to be able to register new items before it loads the data. If it's pre-loaded then such code would need to cause the config to be torn down and re-loaded. I haven't added code for this as yet, as it's not currently a feature that's used, but there were discussions in slack about this being necessary. I'd imagine this would be some re-ordering of imports in `experiment_server.py`.

I added a raise to try and be clearer that if you don't know if something has already had its config loaded you're doing things wrong. I didn't envision people wrapping it in try/except, I imagined that the entry points into the application, either the heroku app or the command line, would be responsible for calling `load_config`. 

Client code would only ever get variables or register new ones, it wouldn't do any of the loading.